Get AoT working

watrool opened the pull request:
Connected to [rangle/rangle-starter#168](https://github.com/rangle/rangle-starter/issues/168)

watrool commented:
Several things to mention:
* The `*.ngfactory.ts` generated by @angular/comipler-cli has syntactical error. So I ignored them during linting.
* After invoking `core-js` in `tsconfig.js` and `package.json` (to solve the error `Cannot find name 'Promise'` during compiling), there are errors about `Duplicate identifier 'PropertyKey'`. And to work around this, I changed the `target` property in `tsconfig.json` from `es6` to `es5`.

@SethDavenport Could you do a review on this?

SethDavenport commented on package.json:
```diff
@@ -11,13 +11,13 @@
     "dev": "cross-env NODE_ENV=development webpack-dev-server -d --inline --hot --progress --no-info",
     "lint": "npm run lint-ts && npm run lint-js && npm run lint-css",
-    "lint-ts": "tslint './src/**/*.ts'",
+    "lint-ts": "tslint './src/**/*.ts' -e './src/**/*.ngfactory.ts'",
```
Makes sense.

SethDavenport commented on package.json:
```diff
@@ -42,16 +42,19 @@
     "@angular/http": "2.0.0-rc.6",
     "@angular/platform-browser": "2.0.0-rc.6",
     "@angular/platform-browser-dynamic": "2.0.0-rc.6",
+    "@angular/platform-server": "2.0.0-rc.6",
```
@barretodavid is platform-server actually needed for AoT?

SethDavenport commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
   "buildOnSave": false,
   "compilerOptions": {
     "module": "commonjs",
```
I seem to recall @bennett000 talking about tree-shaking only working properly with `"module": "es6"`. It would be nice to roll that in here too.

watrool replied:
Setting `"module": "es6" ` breaks the AoT compilation, it leads to errors like `Cannot find module '@angular/router'`. And if set `"target"` also to `"es6"`, `Duplicate identifier 'PropertyKey'` error arises.

SethDavenport commented:
Hey @watrool - so I'm behind on AoT stuff - I'd like someone on the support squad who's more familiar with it to review the approach. @barretodavid @bennett000 @sumitarora who would be a a good person to take a look at this PR?

A couple of things off the top of my head:

__generated output:__

I believe the .ngfactory.ts files are generated output; therefore we probably shouldn't check them into github.  We should put a rule in `.gitignore` to handle this.

__transpilation to es6 or es5:__

Right now our toolchain uses awesome-typescript-loader to transpile to es6, and then babel-loader to transpile to es5. This was done to work around some issues with debugging typescript where source maps would get screwy. @clbond knows the details. It also gives us the ability to use generators and async-await in typescript which sounded very cool but in retrospect I'm not sure anyone actually uses it with Angular's Observable-heavy approach.

It's possible that updating to webpack2 and typescript2 have improved the source map situation. If that's true, then I suggest we try going back to es5 TS output as you have done so we can speed up build time and drop a few dependencies.

Specifically: we no longer need babel, babel-polyfill, babel-loader, etc. (babel-* really).

Give that a shot and see if the following things still work:
* debugging ts code in the browser via source maps
* unit test code coverage reports

watrool commented:
One issue I noticed today: 

Since AoT is statically scanning files, it requires content to be loaded when compiling. So the `@import` directives in  `index.css` cause a problem because its content is not loaded before webpack is called. Then, here comes to a question: should we run a "small" webpack before AoT? (or is there any other way to solve the situation?)

watrool commented on src/app/app.ts:
```diff
@@ -16,7 +16,6 @@ import {BrowserModule} from '@angular/platform-browser';
   selector: 'rio-app',
   // Global styles imported in the app component.
   encapsulation: ViewEncapsulation.None,
-  styleUrls: [ '../styles/index.css' ],
```
After AoT compilation, styleUrls causes problem and stop angular from rendering component.

watrool commented:
AoT + Tree-shaking result:
```

                          Asset       Size  Chunks             Chunk Names
    app.19acebea54aa4ab568f2.js     108 kB       0  [emitted]  app
globals.19acebea54aa4ab568f2.js    33.6 kB       1  [emitted]  globals
 vendor.19acebea54aa4ab568f2.js     431 kB       2  [emitted]  vendor
                     index.html  616 bytes          [emitted]  

```
Before (at https://github.com/rangle/angular2-starter/commit/350219946af14da3dbfb9434f07e52852464fef0):
```
                         Asset       Size  Chunks             Chunk Names
   app.f7d5a963ceb7aec1d21c.js    37.9 kB       0  [emitted]  app
vendor.f7d5a963ceb7aec1d21c.js    1.22 MB       1  [emitted]  vendor
                    index.html  529 bytes          [emitted]  
```

bennett000 commented:
@watrool
> Since AoT is statically scanning files, it requires content to be loaded when compiling. So the @import directives in index.css cause a problem because its content is not loaded before webpack is called. Then, here comes to the question: should we run a "small" webpack before AoT? (or is there any other way to solve the situation?)

Does `require` work in this scenario?

-

> AoT + Tree-shaking result:
```
                          Asset       Size  Chunks             Chunk Names
    app.c0eb0e7bf222f265c88b.js     111 kB       0  [emitted]  app
globals.c0eb0e7bf222f265c88b.js    33.6 kB       1  [emitted]  globals
 vendor.c0eb0e7bf222f265c88b.js     431 kB       2  [emitted]  vendor
                     index.html  616 bytes          [emitted]
```

Wow, that's a huge saings, but 431kB for hello world is still _enormous_

watrool commented:
@bennett000  
I just tried with `require`, and this time AoT is not complaining while webpack rises an error:

```
ERROR in ./src/styles/index.css
Module build failed: Unknown word (1:1)

> 1 | require('basscss-basic');
    | ^
  2 | 
  3 | /* Add-on Modules */
  4 | require( 'basscss-addons/modules/btn');

 @ ./src/app/app.ngfactory.ts 207:13-43
 @ ./src/index.ts
```

So is this a loader error in `webpack.config.js`?

bennett000 commented:
My fault. I'm guessing `require` was deprecated in Webpack 2.

watrool commented:
After updating to RC7, the size is decreased again but the css issue still remains:

```
                          Asset       Size  Chunks             Chunk Names
    app.3cdc79a8873deff8e4fb.js    41.9 kB       0  [emitted]  app
globals.3cdc79a8873deff8e4fb.js    33.6 kB       1  [emitted]  globals
 vendor.3cdc79a8873deff8e4fb.js     422 kB       2  [emitted]  vendor
                     index.html  616 bytes          [emitted]  
```

SethDavenport commented on src/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "sourceMap": false,
+    "inlineSourceMap": true,
+    "noEmitHelpers": true
+  }
```
Why do we have two `.tsconfig.json`s?

SethDavenport commented on package.json:
```diff
@@ -35,23 +35,26 @@
+    "@types/core-js": "^0.9.32",
     "autoprefixer": "^6.3.3",
     "awesome-typescript-loader": "^2.2.1",
     "babel-core": "^6.9.1",
```
If we're getting typescript to emit es5 code, then we can get rid of babel-core, babel-loader, babel-polyfill, and babel-preset-es2015.

watrool replied:
For the build and dev, removing babel-xxx is fine but running karma tests in this case rises problem like: 

`Failed to parse file: /Users/newAccount/Desktop/angular2-starter/src/tests.entry.ts
`

I think the reason for this comes from `istanbul-instrumenter-loader`, mainly because turning off `istanbul-instrumenter-loader` fixes the error (but we loses coverage report). It seems `istanbul-instrumenter-loader` still need babel to transpile ts?

SethDavenport commented:
I just noticed that `dummy-service.test.ts` is in the `styles` folder instead of the `services` folder.

SethDavenport commented:
This is good work though @watrool, interesting results. This is the first time I've seen AoT and tree shaking produce output that's smaller than just using UglifyJS with aggressive settings.

SethDavenport commented:
As for the CSS thing, maybe `@import` isn't the way to go then. Does it work if you `import` into a css file?

You could try doing that in the `app.module.ts`; try applying some basscss classes to `app.html` and see if they stick.

Alternately, you could set them as `styleUrls` in the `@Component` template, with `encapsulation:  ViewEncapsulation.NONE`

watrool commented:
Thanks @SethDavenport , and I have tried both way to enable css, but either `ngc` complains about cannot find resource (because there is no loader?) or `webpack` complains about unexpected character (`import` and `require`). The only successful way is to use `require` and put `require('basscss-basic')`  as a string into the `styles` array in `@component`: but this one is not actually loading css ...

So I guess maybe we need to wait for a newer version of `ngc`.

SethDavenport commented:
We need to have a functional CSS setup before this can be merged. @ericjim do you have any suggestions for doing CSS in Angular with Webpack2 and AoT?

If you use the `styleUrls:` property with `@Component` does it get intercepted by webpack and run through PostCSS?

ericjim commented:
@SethDavenport, if angular-cli supports PostCSS, then it will pick up on it. If it doesn't (which I suspect is the case), then I suggest you add a build step before doing `ng serve` or `ng build --prod`.

For example, earlier in my current project, I was using an earlier version of Angular CLI where SASS was not supported. As a result, I added this in my package.json `"start": "node-sass --output-style compressed --recursive --output ./src/ ./src/ && ngc -p src/ && ng server"`

This line would compile my `.scss` files in-place, and any `@Component`s would pick up on the outputted `.css` file. (in this case `styleUrls:` contains a reference to `.css`)

Note that, this no longer is my setup since SASS is officially supported. Any `.scss` file referenced in `styleUrls:` is automatically picked upon and transpiled to `.css`.

SethDavenport commented:
@ericjim - thanks. Angular-cli supports PostCSS, just not `cssnext`.  PostCSS is the framework they use to transpile CSS from less, and scss. They just don't let us add the cssnext plugin (or autoprefixer for that matter).

@watrool I'm wondering if picking up the styleURLs property and running it through the webpack loaders is something they did custom in the CLI or whether it will work in angular2-starter. Can you try making a simple CSS file with a cssnext construct in it (maybe a variable) and reference it in a component's styleUrls attribute? See what happens?

for example some cssnext like this:

```css
:root {
  --theme-red: #FE0005;
}

h1 {
  color: var(--theme-red);
}
```

watrool commented:
@SethDavenport @ericjim Thanks for your suggestions! 

@SethDavenport I have tried to add cssnext to index.css, however:

1. Using "styleUrls" leads to errors like `Cannot read property 'replace' of undefined` in the console.
2. The reason for the previous error seems to be `encapsulation: ViewEncapsulation.None`, because removing encapsulation fixes the issue.
3. After encapsulation is removed, cssnext does not work. 

Besides, for the original css issue, I think it is a problem of finding resources instead of transpiling css from sass, less or cssnext. The error message is `Resource file not found` and the only walk-around is supply full path of the resource file. For example, change `@import 'basscss-basic'` to `@import '../node_modules/basscss-basic/index.css'`. The reason that this walk-around cannot be applied is that we cannot ask all our dependencies to also use full path.

SethDavenport commented:
OK the node_modules path thing sounds like a config issue with the css loaders or postcss plugins.  @winkerVSbecks this vaguely rings a bell - any tips?

winkerVSbecks commented:
I've seen this issue, but haven't found a workaround yet. Make sure postcss import is included and configured correctly. 

Chris Bond might have some suggestions.

watrool commented:
Thanks @winkerVSbecks ! 
@clbond Could you take a look at this issue?

watrool commented:
Talked with Chris, and tried to adjust postCSS-import config, but still cannot find resource files. I am inclined to think this is a issue with `ngc`.

watrool commented:
Rebased this PR on the latest master and cleaned influenced commits. After this rebase, bundle size become:
```
                         Asset       Size  Chunks             Chunk Names
   app.87bc001a5983d518a493.js     134 kB       0  [emitted]  app
vendor.87bc001a5983d518a493.js     422 kB       1  [emitted]  vendor
                    index.html  537 bytes          [emitted]  
```

One sidenote, when I tried `Rollup` on [Angular AoT page](https://angular.io/docs/ts/latest/cookbook/aot-compiler.html), I got a bundle file with `322` kb.

bennett000 commented:
@watrool this is really interested.  What happens when the app `import`s things like CSS?  If `ngc` runs first will including things like `*.css` files be a problem?

watrool commented:
@bennett000 In my current understanding (`Angular` 2.0.0 with `ngc` 0.6.3), the process is:

1. `ngc` runs first and tries to transpile TS files with respect to `tsconfig.json` (since `ngc` is a wrapper of `tsc`).
2. Due to `stylesUrl` in the `index.ts`, `ngc` starts to locate the css resource, because it wants to

> inlines external html templates and css style sheets within the application JavaScript, eliminating separate ajax requests for those source files.

While it can find `index.css` finely, the `@import 'basscss-addons/modules/btn';` inside `index.css` caused a `Resource file not found` error. It seems `ngc` cannot resolve these css dependencies using the node logic (ie. search `node_modules`). Thus, to solve this, either we can provide it with full path to the resources or we can wait for `ngc` to provide more config options (I believe currently we only have two options in `angularCompilerOptions`).

bennett000 commented:
@watrool thanks for this.  I'm excited that it works.  Unfortunately I don't think pre-building with `ngc` is going to work for us because it breaks webpack for _non_-typescript files.   We might be at the mercy of the `@angular/compiler-cli` team or whatever the `angular-cli` team are doing to work around this issue.

On the positive side, with the `angular-cli` team using webpack we should hopefully see improvements to this at some point (hopefully soon)

watrool commented:
@bennett000 As the issue of `@import 'basscss-basic';` is `Resource file not found`, in https://github.com/rangle/angular2-starter/pull/149/commits/c3e6d031f0010ab20b2cb90d2f699ac7df1f7f56 I wrote an `aot-fix.js` to preload css files before `ngc` transpilation using `postcss`, `postcss-import` and `postcss-cssnext`. And the result is that, now we have a working css.

Besides all the trickiness of this solution, it also prevent the use of `ViewEncapsulation.None` which is raised here: https://github.com/angular/angular/issues/11408.

watrool commented:
After some changes on config, the bundle with css included has:
```
                         Asset       Size  Chunks             Chunk Names
   app.12127d6629679b784046.js      79 kB       0  [emitted]  app
vendor.12127d6629679b784046.js     426 kB       1  [emitted]  vendor
                    index.html  537 bytes          [emitted]  
```
which sums to `505`kb...

watrool commented:
Reduction continues:
```
                         Asset       Size  Chunks             Chunk Names
   app.a7009bf60fe0994ef413.js    55.8 kB       0  [emitted]  app
vendor.a7009bf60fe0994ef413.js     373 kB       1  [emitted]  vendor
                    index.html  537 bytes          [emitted]  
```
which sums to `428.8`kb ...

watrool commented:
@bennett000 What is our expectation of the bundle size?

watrool commented:
The commit https://github.com/rangle/angular2-starter/pull/149/commits/d8284c947589dba845afddf541d73f234b49b13e fixed the error of using `ViewEncapsulation`, and that means we finally get css working on the AoT version.

I think to further imporve this PR, we should also enable AoT as an optional mode for users who wish to use JiT. @SethDavenport and @bennett000 Could you give some suggestions?

watrool commented:
Rebased this PR.

watrool commented:
@smithad15 Andrew I switched this PR to `@ngtools/webpack` and the building is successful with the `aot-fix.js` removed. The result (with CSS included):
```
                         Asset       Size  Chunks             Chunk Names
 0.7bdb831f28449cd179bb.js.map     5.8 kB          [emitted]  
 0.32534de4394153f6584f.js.map    41.9 kB          [emitted]  
   app.4d15f344de2c43b8001d.js    85.3 kB       0  [emitted]  app
vendor.4d15f344de2c43b8001d.js     330 kB       1  [emitted]  vendor
                    index.html  538 bytes          [emitted]  
              assets/README.md   26 bytes          [emitted]  
```
so the size is pretty close to using `ngc` directly. 

Three notes to make though:
1. Using `inlinesourcemap` seems to be prevented with `@ngtools/webpack`. (so I turned to use `soucemap`.
2. Using `ViewEncapsulation.None` causing the `replace of undefined` error in the browser (which is an "old" one and there is a related issue https://github.com/angular/angular/issues/11408).
3. Testing fails in a strange way: `Failed to parse tests.entry.ts` (but I did not change tests.entry.ts, so may need more time on this). @bennett000 Is this the reason we are worried about CI)?

smithad15 commented:
That's a pretty decent bundle size. It looks like you have AOT running for all environments (dev, test, and prod) when we only want it run for production builds. That may clear up a couple of things, like I think we were already doing separate source map files for production, but inline for everything else. 

ViewEncapsulation.None was already on the hit list, so this just confirms the need to load global styles separately. See rangle/rangle-starter#181

Your tests might be fine if you remove the AOT plugin and loader from that environment.

I'll take a look through the rest of the code soon.

smithad15 commented on .babelrc:
```diff
@@ -1,4 +0,0 @@
-{
```
This file is being referenced by `awesome-typescript-loader` as well as `babel-loader` and should be kept

smithad15 commented on package.json:
```diff
@@ -12,12 +12,14 @@
     "prebuild": "npm run clean",
-    "build": "cross-env NODE_ENV=production webpack",
+    "prebuild:jit": "npm run clean",
+    "build": "cross-env NODE_ENV=production webpack --config webpack.config.js",
```
Explicit declaration of config file isn't needed.

smithad15 commented on src/app/app.component.html:
```diff
@@ -1,3 +1,3 @@
-<div>
+<div class="white bg-blue">
```
Can we remove these classes before merge?

smithad15 commented on src/index.html:
```diff
@@ -9,6 +9,8 @@
   </head>
    <body>
-    <rio-app></rio-app>
+    <div>
```
surrounding `<div>` is not required

smithad15 commented on src/main.ts:
```diff
@@ -1,12 +1,9 @@
-import 'babel-polyfill';
-import 'core-js/es6';
-import 'core-js/es7/reflect';
```
Why are we removing these polyfills?

smithad15 commented on tsconfig.json:
```diff
@@ -2,28 +2,23 @@
+    "sourceMap": true,
     "noEmitHelpers": true
   },
-  "awesomeTypescriptLoaderOptions": {
```
AwesomeTypescriptLoader is still required for dev and test

smithad15 commented on package.json:
```diff
@@ -12,12 +12,14 @@
-    "build": "cross-env NODE_ENV=production webpack",
+    "prebuild:jit": "npm run clean",
+    "build": "cross-env NODE_ENV=production webpack --config webpack.config.js",
+    "build:jit": "cross-env NODE_ENV=production webpack",
```
This build step doesn't seem to be disabling AOT as advertised

smithad15 commented on webpack/loaders.js:
```diff
@@ -19,10 +10,7 @@ exports.tslint = {
-  loaders: [
-    'awesome-typescript-loader',
-    'angular2-template-loader',
-  ],
```
These need to be kept for dev and test

smithad15 commented on webpack/loaders.js:
```diff
@@ -1,14 +1,5 @@
-  query: {
-    compact: false,
-  },
-};
```
Was this interfering with AOT? If so, we should remove the package.json dependencies. Beware that `awesome-typescript-loader` is using Babel to go from ES6 to ES5 and may have some cross dependencies.

smithad15 commented on webpack/plugins.js:
```diff
@@ -52,12 +57,8 @@ const prodPlugins = [
-      warnings: false,
-    },
+    beautify: false,
+    comments: false,
```
Reason for the Uglify settings changes? More out of curiosity than anything...

SethDavenport replied:
mangle fnames was broken in one of the angular RCs. But if it's fixed now let's add it back because it helps a lot.

smithad15 commented on src/app/app.component.ts:
```diff
@@ -1,10 +1,8 @@
   templateUrl: './app.component.html',
-  // Global styles imported in the app component.
   styleUrls: ['../styles/index.css'],
-  encapsulation: ViewEncapsulation.None,
```
See rangle/rangle-starter#181 for the continuance of this functionality change.

watrool replied:
Can you show me how to use webpack to load global css into the `<head>` tag? Maybe I can do that in this PR :)

watrool commented:
@smithad15 So it turns out, after angular 2.1.1, we should not use `exclude: /node_modules/` anymore because `@ngtools/webpack` are using `node_modules/@angular` as place to generate `ngfactory.ts` files.

clbond commented:
I don't understand the logic for `exclude: /node_modules/` in the first place. For example if we have a JSON loader, what does it matter if it wants to load a .json file in `src/` or a .json file in `node_modules/`? It's not like the loader is doing a file glob and compiling everything that doesn't match `exclude`. Is there a good reason to have it on any of our loaders?

SethDavenport commented:
No. It also messes up CSS as Katie discovered a couple of weeks ago.

SethDavenport commented:
:shipit:
